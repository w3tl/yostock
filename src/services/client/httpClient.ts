import fetch from "node-fetch";
import { IAPIClient, IResponse } from "./apiClient";

export class HTTPClient implements IAPIClient {
  async get<T>(url: string, params?: Record<string, string>): Promise<IResponse<T>> {
    const callURL: URL = new URL(url);
    if (params) {
      Object.keys(params).forEach(key => callURL.searchParams.append(key, params[key]));
    }

    const response = await fetch(callURL.toString());

    const resp: IResponse<T> = {
      statusCode: response.status,
      result: null,
    };

    resp.result = await response.text();

    return resp;
  }
}