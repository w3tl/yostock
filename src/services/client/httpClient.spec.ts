import { mocked } from 'ts-jest/utils';
import fetch from 'node-fetch';

jest.mock("node-fetch");

import { HTTPClient } from "./httpClient";
const { Response } = jest.requireActual('node-fetch');

const mockedFetch = mocked(fetch).mockImplementation();

describe("HTTP client", () => {
  beforeEach(() => {
    mockedFetch.mockClear();
  });

  it("should receive response", async () => {
    const responseText = "Hello world";
    mockedFetch.mockReturnValueOnce(Promise.resolve(new Response(responseText, {
      status: 200,
    })));

    const client = new HTTPClient();
    const { result } = await client.get("http://example.com");

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(result).toBe(responseText);
  });

  it("should send parameters", async () => {
    mockedFetch.mockReturnValueOnce(Promise.resolve(new Response("ok", {
      status: 200,
    })));

    const client = new HTTPClient();
    await client.get("http://example.com", {
      param1: "value1",
      param2: "value2",
    });

    expect(mockedFetch.mock.calls).toMatchSnapshot();
  });
});