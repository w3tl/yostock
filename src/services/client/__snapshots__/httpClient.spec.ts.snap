// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTTP client should send parameters 1`] = `
Array [
  Array [
    "http://example.com/?param1=value1&param2=value2",
  ],
]
`;
